contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,959 @@
+    /// offers less than 1 OETHb for 1 WETH. In other terms to get 1 OETHB the swap needs to offer 1.0001 WETH
+    /// this is where purchasing OETHb with WETH within the liquidity position is most expensive
+    uint160 public immutable sqrtRatioX96TickLower;
+    /// @notice sqrtRatioX96TickHigher
+    /// @dev tick higher has value 0 and represents 1:1 price parity of WETH to OETHb
+    uint160 public immutable sqrtRatioX96TickHigher;
+    /// @dev tick closest to 1:1 price parity
+    ///      Correctly assesing which tick is closer to 1:1 price parity is important since it affects
+    ///      the way we calculate the underlying assets in check Balance
+    uint160 public immutable sqrtRatioX96TickClosestToParity;
+
+    error NotEnoughWethForSwap(uint256 wethBalance, uint256 requiredWeth); // 0x989e5ca8
+    error NotEnoughWethLiquidity(uint256 wethBalance, uint256 requiredWeth); // 0xa6737d87
+    error PoolRebalanceOutOfBounds(
+        uint256 currentPoolWethShare,
+        uint256 allowedWethShareStart,
+        uint256 allowedWethShareEnd
+    ); // 0x3681e8e0
+    error OutsideExpectedTickRange(int24 currentTick); // 0x46a58db6
+
+    event PoolRebalanced(uint256 currentPoolWethShare);
+
+    event PoolWethShareIntervalUpdated(
+        uint256 allowedWethShareStart,
+        uint256 allowedWethShareEnd
+    );
+
+    event PrincipalPositionAfterSwap(
+        uint256 wethPositionBalance,
+        uint256 oethbPositionBalance
+    );
+    event LiquidityRemoved(
+        uint256 withdrawLiquidityShare,
+        uint256 removedWETHAmount,
+        uint256 removedOETHbAmount,
+        uint256 wethAmountCollected,
+        uint256 oethbAmountCollected,
+        uint256 underlyingAssets
+    );
+
+    event LiquidityAdded(
+        uint256 wethAmountDesired,
+        uint256 oethbAmountDesired,
+        uint256 wethAmountSupplied,
+        uint256 oethbAmountSupplied,
+        uint256 tokenId,
+        uint256 underlyingAssets
+    );
+
+    event UnderlyingAssetsUpdated(uint256 underlyingAssets);
+
+    /**
+     * @dev Verifies that the caller is the Governor, or Strategist.
+     */
+    modifier onlyGovernorOrStrategist() {
+        require(
+            msg.sender == governor() ||
+                msg.sender == IVault(vaultAddress).strategistAddr(),
+            "Not the Governor or Strategist"
+        );
+        _;
+    }
+
+    /**
+     * @dev Un-stakes the token from the gauge for the execution duration of
+     * the function and after that re-stakes it back in.
+     *
+     * It is important that the token is unstaked and owned by the strategy contract
+     * during any liquidity altering operations and that it is re-staked back into the
+     * gauge after liquidity changes. If the token fails to re-stake back to the
+     * gauge it is not earning incentives.
+     */
+    modifier gaugeUnstakeAndRestake() {
+        if (tokenId != 0 && lpTokenDepositedToGauge) {
```

Comment:
Lol, you probably don't want to do this, and maybe it's a bad idea, but if we always left one liquidity in the token after we create it, then depositing works 100% of the time.

Reply:
It's really all the complex around managing these booleans and checking them. It's error prone - and in fact there's already an error in the code around them.

The bools aren't a gas savings either - it's four storage writes in a typical rebalance.

What if instead we just made an internal function that asks the NFT contract if it's on the gauge, then used that. We still have to check it correctly when using the gauge, but it's always going to be right.

Reply:
> What if instead we just made an internal function that asks the NFT contract if it's on the gauge, then used that. We still have to check it correctly when using the gauge, but it's always going to be right.

+1 to this. It could be as simple as just checking the owner of the NFT token (`ownerOf(tokenId) != address(this)` perhaps?)

Reply:
Agree with both and a great proposal on how to implement the solution. Clears up state handling of the code quite a bit. Fix here: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/b30756ad7e0da7d724372ebee144ca4b9fe5ea10